### ISSUE TITLE
Misleading diagnostics for `.iter()`

### ISSUE THREAD

**Issue by avandesa** [rust-lang/rust#68445]
The following code:
```rust
fn main() {
    let iter = vec![1, 2, 3].iter().map(|x| x * x);

    for x in iter {
        println!("{}", x);
    }
}
```

Produces the following error:
```
error[E0716]: temporary value dropped while borrowed
 --> foo.rs:2:16
  |
2 |     let iter = vec![1, 2, 3].iter().map(|x| x * x);
  |                ^^^^^^^^^^^^^                      - temporary value is freed at the end of this statement
  |                |
  |                creates a temporary which is freed while still in use
3 | 
4 |     for x in iter {
  |              ---- borrow later used here
  |
  = note: consider using a `let` binding to create a longer lived value
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0716`.
```

Though the suggestion to make a let binding works, and is probably useful for other situations, it's not as elegant here:
```rust
fn main() {
    let vec = complicated_get_vec();
    let iter = vec.iter().map(|x| x * x);

    for x in iter {
        println!("{}", x);
    }
}
```

The better solution to the compile error is to replace `vec![].iter() ...` with `vec![].into_iter()`, and the compiler error message may be updated to reflect this.

**Comment by estebank** [comment]
Current output:

```
error[E0716]: temporary value dropped while borrowed
 --> src/main.rs:2:16
  |
2 |     let iter = vec![1, 2, 3].iter().map(|x| x * x);
  |                ^^^^^^^^^^^^^                      - temporary value is freed at the end of this statement
  |                |
  |                creates a temporary value which is freed while still in use
3 |
4 |     for x in iter {
  |              ---- borrow later used here
  |
  = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider using a `let` binding to create a longer lived value
  |
2 ~     let binding = vec![1, 2, 3];
3 ~     let iter = binding.iter().map(|x| x * x);
  |
```

We still don't look if we could suggest `into_iter()` instead.
